### PR TITLE
Issue #1731 - Add support for ':not' modifier

### PR DESCRIPTION
**Pull request by tbieste**
Added support for the search parameter ':not' modifier.

Signed-off-by: Troy Biesterfeld <tbieste@us.ibm.com>

### CODE REVIEW

File: docs/src/pages/Conformance.md [hunk 1]
```diff
@@ -167,9 +165,9 @@ FHIR search modifiers are described at https://www.hl7.org/fhir/R4/search.html#m
 
 Due to performance implications, the `:exact` modifier should be used for String searches where possible.
 
-At present, the `:missing` modifier is not supported for whole-system search nor for chained parameter search. For example, a search with query string like `subject:Basic.date:missing` will result in an `OperationOutcome` explaining that the search parameter could not be processed.
+At present, the `:missing` and `:not` modifiers are not supported for whole-system search nor for chained parameter search. For example, a search with query string like `subject:Basic.date:missing` will result in an `OperationOutcome` explaining that the search parameter could not be processed.
 
-The :text, :not, :above, :below, :in, :not-in, and :of-type modifiers are not supported in this version of the IBM FHIR server and use of this modifier will results in an HTTP 400 error with an OperationOutcome that describes the failure.
+The :text, :above, :below, :in, :not-in, and :of-type modifiers are not supported in this version of the IBM FHIR server and use of this modifier will results in an HTTP 400 error with an OperationOutcome that describes the failure.
```
Comment (michaelwschroeder) [inline docs/src/pages/Conformance.md:170]:
```suggestion
The :text, :above, :below, :in, :not-in, and :of-type modifiers are not supported in this version of the IBM FHIR server and use of these modifiers will result in an HTTP 400 error with an OperationOutcome that describes the failure.
```

File: docs/src/pages/Conformance.md [hunk 2]
```diff
@@ -167,9 +165,9 @@ FHIR search modifiers are described at https://www.hl7.org/fhir/R4/search.html#m
 
 Due to performance implications, the `:exact` modifier should be used for String searches where possible.
 
-At present, the `:missing` modifier is not supported for whole-system search nor for chained parameter search. For example, a search with query string like `subject:Basic.date:missing` will result in an `OperationOutcome` explaining that the search parameter could not be processed.
+At present, the `:missing` and `:not` modifiers are not supported for whole-system search nor for chained parameter search. For example, a search with query string like `subject:Basic.date:missing` will result in an `OperationOutcome` explaining that the search parameter could not be processed.
```
Comment (michaelwschroeder) [inline docs/src/pages/Conformance.md:168]:
```suggestion
At present, the `:missing` and `:not` modifiers are not supported for whole-system search nor for chained parameter search. For example, a search with a query string like `subject:Basic.date:missing` will result in an `OperationOutcome` explaining that the search parameter could not be processed.
```

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java [hunk 3]
```diff
@@ -753,23 +753,24 @@ private boolean isSystemLevelSearch(Class<? extends Resource> resourceType) {
...
                             + "for chained parameters [code=" + param.getCode() + "]");
                 }
             }
 
             do {
                 if (param.getModifier() != null &&
                         !JDBCConstants.supportedModifiersMap.get(param.getType()).contains(param.getModifier())) {
-                    throw buildNotSupportedException("Found unsupported modifier '" + param.getModifier() + "'"
+                    throw buildNotSupportedException("Found unsupported modifier ':" + param.getModifier().value() + "'"
                             + " for search parameter '" + param.getCode() + "' of type " + param.getType());
```
Comment (michaelwschroeder) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:774]:
```suggestion
                            + " for search parameter '" + param.getCode() + "' of type '" + param.getType() + "'");
```

**Review by michaelwschroeder — APPROVED** [review]
LGTM - just some minor doc comments

File: docs/src/pages/Conformance.md (hunk 1 above)
Comment (tbieste) [reply docs/src/pages/Conformance.md:170]: Updated.

File: docs/src/pages/Conformance.md (hunk 2 above)
Comment (tbieste) [reply docs/src/pages/Conformance.md:168]: Updated.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java (hunk 3 above)
Comment (tbieste) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java]: Updated.

**Review by lmsurpre — APPROVED** [review]
LGTM

**Review by punktilious — APPROVED** [review]
LGTM, but would like to see some full queries and check their plans.

**Comment by tbieste** [comment]
> LGTM, but would like to see some full queries and check their plans.

Here's the most common query I was using.

Search query:
Patient?gender:not=male

SQL count query:
SELECT COUNT(DISTINCT R.LOGICAL_RESOURCE_ID)  FROM Patient_LOGICAL_RESOURCES LR  JOIN Patient_RESOURCES R ON R.RESOURCE_ID = LR.CURRENT_RESOURCE_ID  AND R.IS_DELETED = 'N'  WHERE NOT EXISTS (SELECT 1 FROM Patient_TOKEN_VALUES_V  AS param0 WHERE (param0.PARAMETER_NAME_ID = 1180 AND ((param0.TOKEN_VALUE = ?))) AND LR.LOGICAL_RESOURCE_ID = param0.LOGICAL_RESOURCE_ID)

SQL results query:
SELECT R.RESOURCE_ID, R.LOGICAL_RESOURCE_ID, R.VERSION_ID, R.LAST_UPDATED, R.IS_DELETED, R.DATA, LR.LOGICAL_ID  FROM PATIENT_RESOURCES R JOIN (SELECT DISTINCT LR.LOGICAL_RESOURCE_ID, LR.LOGICAL_ID, LR.CURRENT_RESOURCE_ID FROM Patient_LOGICAL_RESOURCES LR  JOIN Patient_RESOURCES R ON R.RESOURCE_ID = LR.CURRENT_RESOURCE_ID  AND R.IS_DELETED = 'N'  WHERE NOT EXISTS (SELECT 1 FROM Patient_TOKEN_VALUES_V  AS param0 WHERE (param0.PARAMETER_NAME_ID = 1180 AND ((param0.TOKEN_VALUE = ?))) AND LR.LOGICAL_RESOURCE_ID = param0.LOGICAL_RESOURCE_ID)) AS LR  ON  R.LOGICAL_RESOURCE_ID = LR.LOGICAL_RESOURCE_ID  AND R.RESOURCE_ID = LR.CURRENT_RESOURCE_ID  AND R.IS_DELETED = 'N' ORDER BY LOGICAL_RESOURCE_ID ASC  OFFSET 0 ROWS  FETCH NEXT 10 ROWS ONLY

SQL binding variable:
[? = 'male']

**Comment by tbieste** [comment]
Comments addressed. Ready to merge.

**Pull request merged**